internal/discovery/dkv_discovery_service.go

```diff
@@ -0,0 +1,130 @@
+	// after exceeding this time the region & node combination can be marked invalid
+	heartbeatTimeout	uint64
+}
+
+func NewDiscoverConfigFromIni(sect *ini.Section) *DiscoveryConfig {
+	sectConf := sect.KeysHash()
+	if statusTTL, err := strconv.ParseUint(sectConf["statusTTL"], 10, 64); err == nil {
+		if heartbeatTimeout, err := strconv.ParseUint(sectConf["heartbeatTimeout"], 10, 64); err == nil {
+			return NewDiscoverConfig(statusTTL, heartbeatTimeout)
+		}
+	}
+	panic(fmt.Errorf("Invalid discovery server configuration. Check section %s", sect.Name()))
+}
+
+func NewDiscoverConfig(statusTTL uint64, heartbeatTimeout uint64) *DiscoveryConfig {
+	return &DiscoveryConfig{statusTTL, heartbeatTimeout}
+}
+
+type discoverService struct {
+	// TODO - dkvServer should point to interface of master (currently distributedService)
+	dkvServer DKVServer
+	logger    *zap.Logger
+	config    *DiscoveryConfig
+}
+
+// Create the dkv lookup key from region info
+// Since granularity of an update is at a region per node, the same is used for creating a key
+// As lookups will be mostly on database and possibly on vBucket (within a database) the same order is used for key
+// This enables faster scans for database/vBucket
+func createKeyToInsert(info *RegionInfo) []byte {
+	return []byte(info.GetDatabase() + ":" + info.GetVBucket() + ":" + info.GetNodeAddress())
+}
+
+// Create lookup key based on input request
+// Using the sorted nature of keys, lookup by either database or by database and vBucket or all keys
+func createKeyToGet(request *GetClusterInfoRequest) []byte {
+	if request.GetDatabase() != "" {
+		if request.GetVBucket() != "" {
+			return []byte(request.GetDatabase() + ":" + request.GetVBucket() + ":");
+		} else {
+			return []byte(request.GetDatabase() + ":");
+		}
+	} else {
+		return []byte{};
+	}
+}
+
+func (d *discoverService) UpdateStatus(ctx Context, request *UpdateStatusRequest) (*Status, error) {
+	// Store the serialised form of the request
+	val, err := json.Marshal(request);
+
+	if err != nil {
+		d.logger.Error("Unable to marshal status request", zap.Error(err));
+		return newErrorStatus(err), err;
+	}
+
+	// TODO - check if expiry TTL is working
+	putRequest := PutRequest{Key: createKeyToInsert(request.GetRegionInfo()), Value: val, ExpireTS: hlc.GetUnixTimeFromNow(d.config.statusTTl)};
+
+	if _, err := d.dkvServer.Put(ctx, &putRequest); err != nil {
+		return newErrorStatus(err), err;
+	} else {
+		return newEmptyStatus(), nil;
+	}
+}
+
+func (d *discoverService) GetClusterInfo(ctx Context, request *GetClusterInfoRequest) (*GetClusterInfoResponse, error) {
+	// Create a prefix multi get request to loop through all required status updates
+	// Iterator sends the responses on grpc stream which is unnecessary here
+	getRequest := PrefixMultiGetRequest{KeyPrefix: createKeyToGet(request)}
+
+	if clusterInfo, err := d.dkvServer.PrefixMultiGet(ctx, &getRequest); err == nil {
```

Comment:
Grpc uses http2, hence no new channel is created. A normal rpc would lead to a single response vs a streaming one would send multiple responses. IMO we can avoid duplicate code by reusing existing constructs.

Reply:
Fixed by Moving to Iterate #81